homeassistant/components/zigate.py

```diff
@@ -0,0 +1,278 @@
+RAW_COMMAND_SCHEMA = vol.Schema({
+    vol.Required('cmd'): cv.positive_int,
+    vol.Optional('data'): cv.string,
+})
+
+IDENTIFY_SCHEMA = vol.Schema({
+    vol.Required(ADDR): cv.string,
+})
+
+
+def setup(hass, config):
+    """Setup zigate platform."""
+    from homeassistant.components import persistent_notification
+    import zigate
+
+    port = config.get(CONF_PORT)
+    persistent_file = os.path.join(hass.config.config_dir,
+                                   'zigate.json')
+    _LOGGER.debug('Persistent file {}'.format(persistent_file))
+
+    z = zigate.ZiGate(port,
+                      persistent_file,
+                      auto_start=False)
+
+    hass.data[DOMAIN] = z
+    hass.data[DATA_ZIGATE_DEVICES] = {}
+    hass.data[DATA_ZIGATE_ATTRS] = {}
+
+    component = EntityComponent(_LOGGER, DOMAIN, hass)
+
+    def device_added(**kwargs):
+        device = kwargs['device']
+        _LOGGER.debug('Add device {}'.format(device))
+        if device.addr not in hass.data[DATA_ZIGATE_DEVICES]:
+            entity = ZiGateDeviceEntity(device)
+            hass.data[DATA_ZIGATE_DEVICES][device.addr] = entity
+            component.add_entities([entity])
+            if 'signal' in kwargs:
+                persistent_notification.create(hass,
+                                               ('A new ZiGate device "{}"'
+                                                ' has been added !'
+                                                ).format(device),
+                                               title='ZiGate')
+
+    def device_removed(**kwargs):
+        # component.async_remove_entity
+        pass
+
+    def device_need_refresh(**kwargs):
+        device = kwargs['device']
+        persistent_notification.create(hass,
+                                       ('The ZiGate device {} needs some'
+                                        ' refresh (missing important'
+                                        ' information)').format(device.addr),
+                                       title='ZiGate')
+
+    zigate.dispatcher.connect(device_added,
+                              zigate.ZIGATE_DEVICE_ADDED, weak=False)
+    zigate.dispatcher.connect(device_removed,
+                              zigate.ZIGATE_DEVICE_REMOVED, weak=False)
+    zigate.dispatcher.connect(device_need_refresh,
+                              zigate.ZIGATE_DEVICE_NEED_REFRESH, weak=False)
+
+    def attribute_updated(**kwargs):
+        device = kwargs['device']
+        attribute = kwargs['attribute']
+        _LOGGER.debug('Update attribute for device {} {}'.format(device,
+                                                                 attribute))
+        key = '{}-{}-{}-{}'.format(device.addr,
+                                   attribute['endpoint'],
+                                   attribute['cluster'],
+                                   attribute['attribute'],
+                                   )
+        entity = hass.data[DATA_ZIGATE_ATTRS].get(key)
```

Comment:
Platform entities should not be accessed directly outside their platforms. Entities shouldn't be stored under a shared key in hass.data. It's ok to store a reference to the entity eg the entity_id under a shared key.

If the component needs to signal the entity, eg to update state, that should be done via our dispatch helper.

See the recently added tuya component for example.